conformance/03_discovery_test.go

```diff
@@ -82,9 +84,151 @@ var test03ContentDiscovery = func() {
 				RunOnlyIfNot(runContentDiscoverySetup)
 				tagList = strings.Split(os.Getenv(envVarTagList), ",")
 			})
+
+			g.Specify("References setup", func() {
+				SkipIfDisabled(contentDiscovery)
+				RunOnlyIf(runContentDiscoverySetup)
+
+				// Populate registry with empty JSON blob
+				// validate expected empty JSON blob digest
+				Expect(emptyJSONDescriptor.Digest).To(Equal(godigest.Digest("sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a")))
+				req := client.NewRequest(reggie.POST, "/v2/<name>/blobs/uploads/")
+				resp, err := client.Do(req)
+				Expect(err).To(BeNil())
+				req = client.NewRequest(reggie.PUT, resp.GetRelativeLocation()).
+					SetQueryParam("digest", emptyJSONDescriptor.Digest.String()).
+					SetHeader("Content-Type", "application/octet-stream").
+					SetHeader("Content-Length", fmt.Sprintf("%d", emptyJSONDescriptor.Size)).
+					SetBody(emptyJSONBlob)
+				resp, err = client.Do(req)
+				Expect(err).To(BeNil())
+				Expect(resp.StatusCode()).To(SatisfyAll(
+					BeNumerically(">=", 200),
+					BeNumerically("<", 300)))
+
+				// Populate registry with reference blob before the image manifest is pushed
+				req = client.NewRequest(reggie.POST, "/v2/<name>/blobs/uploads/")
+				resp, err = client.Do(req)
+				Expect(err).To(BeNil())
+				req = client.NewRequest(reggie.PUT, resp.GetRelativeLocation()).
+					SetQueryParam("digest", testRefBlobADigest).
+					SetHeader("Content-Type", "application/octet-stream").
+					SetHeader("Content-Length", testRefBlobALength).
+					SetBody(testRefBlobA)
+				resp, err = client.Do(req)
+				Expect(err).To(BeNil())
+				Expect(resp.StatusCode()).To(SatisfyAll(
+					BeNumerically(">=", 200),
+					BeNumerically("<", 300)))
+
+				// Populate registry with test references manifest (config.MediaType = artifactType)
+				req = client.NewRequest(reggie.PUT, "/v2/<name>/manifests/<reference>",
+					reggie.WithReference(refsManifestAConfigArtifactDigest)).
+					SetHeader("Content-Type", "application/vnd.oci.image.manifest.v1+json").
+					SetBody(refsManifestAConfigArtifactContent)
+				resp, err = client.Do(req)
+				Expect(err).To(BeNil())
+				Expect(resp.StatusCode()).To(SatisfyAll(
+					BeNumerically(">=", 200),
+					BeNumerically("<", 300)))
+				Expect(resp.Header().Get("OCI-Subject")).To(Equal(manifests[4].Digest))
+
+				// Populate registry with test references manifest (ArtifactType, config.MediaType = emptyJSON)
+				req = client.NewRequest(reggie.PUT, "/v2/<name>/manifests/<reference>",
+					reggie.WithReference(refsManifestALayerArtifactDigest)).
+					SetHeader("Content-Type", "application/vnd.oci.image.manifest.v1+json").
+					SetBody(refsManifestALayerArtifactContent)
```

Comment:
Hi @jdolitsky

Looking at this test @peusebiu and I noticed `refsManifestALayerArtifactContent` actually contains a reference to layer `testRefBlobB` (see https://github.com/opencontainers/distribution-spec/blob/main/conformance/setup.go#L396), which is not yet uploaded to the registry.

My understanding of the push specifications is that layer blobs should be pushed by the client before the manifest.
If the server decides to reject them, it remains fully compliant. So why would the test check the server accepts this upload.

We're not sure if this is by design of the test, or a bug, as `testRefBlobB` is uploaded later at line 198.
On the other hand the code in setup.go is strange, maybe `testRefBlobA` should have been reference there is setup.go at line 396

Reply:
Yes, I defer to Ram. In this PR I tried to just copy-paste his work